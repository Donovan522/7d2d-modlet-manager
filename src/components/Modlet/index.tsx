import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import TableRow from "@material-ui/core/TableRow";
import TableCell from "@material-ui/core/TableCell";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import Switch from "@material-ui/core/Switch";
import Typography from "@material-ui/core/Typography";
import { Modlet } from "helpers";
import PropTypes from "prop-types";
import React, { useState } from "react";

interface ModletProps {
  advancedMode: boolean;
  modlet: Modlet;
}

function ModletComponent(props: ModletProps): React.ReactElement {
  const conditions: React.ReactNode[] = [];

  const [enabled, setEnabled] = useState(props.modlet.isEnabled());

  const handleEnableClick = (event: React.ChangeEvent<HTMLInputElement>, modlet: Modlet) => {
    setEnabled(event.target.checked);
    modlet.enable(event.target.checked);
  };

  const enableSwitch: React.ReactNode = (
    <FormControlLabel
      style={{ marginLeft: "auto", marginRight: 5 }}
      control={<Switch size="small" checked={enabled} onChange={e => handleEnableClick(e, props.modlet)} />}
      label={enabled ? "Enabled" : "Disabled"}
      labelPlacement="start"
    />
  );

  const compatability = props.modlet.get("compat");

  return props.advancedMode ? (
    <TableRow>
      <TableCell>
        <Typography variant="body1">{props.modlet.get("name")}</Typography>
        <Typography variant="body2" color="textSecondary">
          {props.modlet.get("description")}
        </Typography>
      </TableCell>
      <TableCell align="right">
        <Typography variant="body1">{props.modlet.get("version")}</Typography>
        <Typography variant="body2" color="textSecondary">
          {compatability}
        </Typography>
      </TableCell>
      <TableCell>{conditions}</TableCell>
    </TableRow>
  ) : (
    <Card>
      <CardContent>
        <Typography variant="body1">{props.modlet.get("name")}</Typography>
        <Typography variant="body2" color="textSecondary">
          {props.modlet.get("description")}
        </Typography>
        <Typography style={{ paddingTop: 15 }} variant="caption" component="div" color="textSecondary">
          By {props.modlet.get("author")} - v.{props.modlet.get("version")}&nbsp;
          {props.modlet.get("compat") !== "unknown" && <i>(compatible with: {props.modlet.get("compat")})</i>}
        </Typography>
      </CardContent>
      <CardActions>{enableSwitch}</CardActions>
    </Card>
  );
}

ModletComponent.propTypes = {
  advancedMode: PropTypes.bool.isRequired,
  modlet: PropTypes.instanceOf(Modlet).isRequired
};

export default ModletComponent;
