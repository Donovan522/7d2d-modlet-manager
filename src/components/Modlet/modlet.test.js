import React from "react";
import Modlet from ".";
import createMockModlet from "test_helpers/mock_modlet";
import mock_state from "test_helpers/mock_state";
import { render } from "@testing-library/react";

let modletState;
const table = document.createElement("table");

describe("Modlet", () => {
  beforeEach(() => {
    modletState = {
      modlet: createMockModlet(),
      validated: false,
      errors: []
    };
  });

  describe("Basic Mode", () => {
    beforeEach(() => {
      mock_state.advancedMode = false;
    });

    it("should render correctly", () => {
      const { asFragment } = render(
        <Modlet state={mock_state} modletState={modletState} handleValidation={jest.fn()} />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("Advanced Mode", () => {
    beforeEach(() => {
      mock_state.advancedMode = true;
    });

    it("should render correctly", () => {
      const { asFragment } = render(
        <Modlet state={mock_state} modletState={modletState} handleValidation={jest.fn()} />,
        {
          container: document.body.appendChild(table),
          wrapper: "tbody"
        }
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
