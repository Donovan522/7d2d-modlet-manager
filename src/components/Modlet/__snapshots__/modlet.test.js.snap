// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modlet Advanced Mode should render correctly 1`] = `
<DocumentFragment>
  <tbody>
    <tr
      class="MuiTableRow-root"
    >
      <td
        class="MuiTableCell-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          test name
        </p>
        <p
          class="MuiTypography-root makeStyles-description-121 MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          test description
        </p>
      </td>
      <td
        class="MuiTableCell-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          test author
        </p>
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-alignRight"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
        >
          normal 1.0.1
        </p>
        <p
          class="MuiTypography-root makeStyles-italic-120 MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          unknown
        </p>
      </td>
      <td
        class="MuiTableCell-root"
      >
        <div
          class="MuiBox-root MuiBox-root-174 makeStyles-box-117"
        >
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-187 MuiCheckbox-root MuiCheckbox-colorSecondary makeStyles-checkbox-124 PrivateSwitchBase-checked-188 Mui-checked MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-190"
                  data-indeterminate="false"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-iconOK-122"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Installed
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-187 MuiCheckbox-root MuiCheckbox-colorSecondary makeStyles-checkbox-124 PrivateSwitchBase-checked-188 Mui-checked MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  checked=""
                  class="PrivateSwitchBase-input-190"
                  data-indeterminate="false"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root makeStyles-iconOK-122"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
              </span>
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Enabled
            </span>
          </label>
          <label
            class="MuiFormControlLabel-root"
          >
            <span
              aria-disabled="false"
              class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-187 MuiCheckbox-root MuiCheckbox-colorSecondary makeStyles-checkbox-124 MuiIconButton-colorSecondary"
            >
              <span
                class="MuiIconButton-label"
              >
                <input
                  class="PrivateSwitchBase-input-190"
                  data-indeterminate="false"
                  type="checkbox"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                  />
                </svg>
              </span>
            </span>
            <span
              class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
            >
              Validate
            </span>
          </label>
        </div>
      </td>
    </tr>
  </tbody>
</DocumentFragment>
`;

exports[`Modlet Basic Mode should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-root-1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        test name
      </p>
      <p
        class="MuiTypography-root makeStyles-description-6 MuiTypography-body2 MuiTypography-colorTextSecondary"
      >
        test description
      </p>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing"
    >
      <div
        class="MuiTypography-root makeStyles-cardVersion-3 MuiTypography-caption MuiTypography-colorTextSecondary"
      >
        By test author - vnormal 1.0.1 
      </div>
      <label
        class="MuiFormControlLabel-root makeStyles-enableControl-4 MuiFormControlLabel-labelPlacementStart"
      >
        <span
          class="MuiSwitch-root MuiSwitch-sizeSmall"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-90 MuiSwitch-switchBase MuiSwitch-colorSecondary PrivateSwitchBase-checked-91 Mui-checked"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input-93 MuiSwitch-input"
                type="checkbox"
                value=""
              />
              <span
                class="MuiSwitch-thumb"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Enabled
        </span>
      </label>
    </div>
  </div>
</DocumentFragment>
`;
